instructors.md

```diff
@@ -90,18 +90,395 @@ which are covered in the lesson on the shell.
+>
+> ~~~ {.python}
+> for number in range(10):
+>     # use a if the number is a multiple of 3, otherwise use b
+>     if (Number % 3) == 0:
+>         message = message + a
+>     else:
+>         message = message + "b"
+> print message
+> ~~~
+> Answer: 3 `NameError`s for `number` being misspelled, for `message` not defined, and for `a` not being in quotes.  
+> ~~~ {.python}
+> message = ""
+> for number in range(10):
+>     # use a if the number is a multiple of 3, otherwise use b
+>     if (Number % 3) == 0:
+>         message = message + "a"
+>     else:
+>         message = message + "b"
+> print message
+> abbabbabba
+
+> ## Identifying Item Errors {.challenge}
+>
+> 1. Read the code below, and (without running it) try to identify what the errors are.
+> 2. Run the code, and read the error message. What type of error is it?
+> 3. Fix the error.
+>
+> ~~~ {.python}
+> seasons = ['Spring', 'Summer', 'Fall', 'Winter']
+> print 'My favorite season is ', seasons[4]
+> ~~~
+> Answer: IndexError; the last entry is `seasons[3]`, so `seasons[4]` doesn't make sense.
+> ~~~ {.python}
+> seasons = ['Spring', 'Summer', 'Fall', 'Winter']
+> print 'My favorite season is ', seasons[-1]
+> ~~~
+
 ## [Defensive Programming](08-defensive.html)
 
+Solutions to exercises:
+> ## Pre- and post-conditions {.challenge}
+>
+> Suppose you are writing a function called `average` that calculates the average of the numbers in a list.
+> What pre-conditions and post-conditions would you write for it?
+> Compare your answer to your neighbor's:
+> can you think of a function that will pass your tests but not hers or vice versa?
+> Answer: 
+> possible pre-conditions: 
+> ~~~ {.python}
+> assert len(input)>0, 'List length must be non-zero'
+> import numbers, numpy
+> for n in input_list:
+>     assert isinstance(n, numbers.Number) and not numpy.isnan(n),  'list entries must be numbers'
+> ~~~
+> possible post-conditions: 
+> ~~~ {.python}
+> assert not numpy.isnan(average), 'Average computation failed'
+> ~~~
+
+> ## Testing assertions {.challenge}
+>
+> Given a sequence of values, the function `running` returns
+> a list containing the running totals at each index.
+> Explain in words what the assertions in this function check,
+> and for each one,
+> give an example of input that will make that assertion fail.
+>
+> ~~~ {.python}
+> def running(values):
+>     assert len(values) > 0
+>     result = [values[0]]
+>     for v in values[1:]:
+>         assert result[-1] >= 0
+>         result.append(result[-1] + v)
+>         assert result[-1] >= result[0]
+>     return result
+> ~~~
+> Answer:
+> The first assertion checks that the input sequence `values` is not empty; an empty sequence such as `[]` will make it fail.
+> The second assertion checks that the last item in `result` exists (FIXME: not obvious to me how to make this fail)  
+> The third assertion checks that the last item in `result` exists (FIXME: not obvious to me how to make this fail)  
```

Comment:
This tests that your sum is always greater than the first value. For instance this fails:
values = range(5, 10)
values.append(-34)
print running(values)